### ISSUE TITLE
Add support for Webfinger - RFC 7033

### ISSUE THREAD

**Issue by edrf12** [kanidm/kanidm#3162]
### Why Webfinger

Webfinger is a pretty useful thing for other services that want to discover information about some URI, one such services is [Tailscale](https://tailscale.com). Since Kanidm supports OIDC Webfinger would allow services to easily discover the configuration for verifying a user.

### What could be done

Kanidm could have a Webfinger specific public client that could either be enabled or disabled. This would indeed conflict with the philosophy of having one client id for each application, but I think It's an interesting compromise for some services and could be circumvented with scopes.

### Current Solution

I've been maintaining a secondary service [Maronato/go-finger](https://github.com/Maronato/go-finger) but having it directly supported in Kanidm makes things much simpler and exclude the need of having two edited configurations

### Additional context

https://datatracker.ietf.org/doc/html/rfc7033
https://webfinger.net/

**Comment by yaleman** [comment]
We would need to have a "primary" response with one app per domain, or enforce a requirement that scopes be specified which from memory Tailscale doesn't support... also the fact that they use email (which we classify as PII) as an identifier is problematic.

**Comment by Firstyear** [comment]
The problem here is that there is a single url for web-finger, but we support *multiple* clients. So we'd need a way to direct someone to the correct public web-fingered client, or rather a way for the admin to say 'this is the client people will go to after a web-finger request'. I'm not sure how we express that properly - either a built-in and default-disabled public client that is bound to web-finger, or we have a marker on the oauth2 client definition that denotes it as the web-finger recipient, or a different config in the db that links to the active oauth2 client.

**Comment by CEbbinghaus** [comment]
This could be done the same way Nextcloud does. By implementing the webfinger support such that it is under the same path as OIDC Discovery url `https://idm.example.com/oauth2/openid/:client_id:/.well-known/webfinger`, we give the user the power to decide for themselves how they want it set up. For example they could designate one of their OIDC clients as "Primary" and just redirect every request to `https://hostname.tld/.well-known/webfinger` to `https://idm.example.com/oauth2/openid/primary/.well-known/webfinger` (either via 301 or through url rewriting in the reverse proxy). Another option would be to seprate clients into subdomains and just have `https://gitlab.hostname.tld/.well-known/webfinger` redirect to the `gitlab` client webfinger url.

I suspect we would want to offer some guidance on good ways to set it up but at the end of the day by implementing it as a generic solution we allow the admin to set up webfinger however they want. The only downside to this being that using webfinger would require a reverse proxy which at this point is not exactly a huge ask (since most run it already). But perhaps considering a db entry to assign a "default" client (as @Firstyear suggested) might still be desired depending on the amount of people wanting webfinger that do not run a reverse proxy.

Would be happy to take on this work :3

**Comment by CEbbinghaus** [comment]
Just to expand on the above, As @yaleman pointed out in the community chat the email is considered PII and therefore we cannot deny a request based on whether the email requested within our database. As the email host must match our kanidm hostname it would make more sense to use the spn instead. This does mean that it is not guaranteed to be a valid email target but without it we would be entirely unable to log in to begin with (given the email wouldn't match our hostname). However it does require that the users spn is provided as email otherwise the service would throw an error due to the emails not matching (if the service does things right).

It would likely require pointing out this behavior in the docs if we do implement it given its implications, It does however add to the list of wierd quirks that makes this feature less desirable. But given at times this is the only way to proceed with using kanidm as OIDC (e.g Tailscale)I think its still worthwhile to seriously consider. Despite the wierdness and "magic"

**Comment by Firstyear** [comment]
Okay, lets start with the per client webfinger then, and then we can do the "default" validation in a subsequent PR. Are you happy to take this on then @CEbbinghaus ?

**Comment by CEbbinghaus** [comment]
Yup. I'll do it. First time following an RFC but this one seems simple enough
